File: apps/fintech/before/lib/store.ts
export type Account = {
  id: string;
  customerId: string;
  type: "checking" | "savings" | "credit";
  balance: number;
  currency: string;
  status: "active" | "frozen" | "closed";
  createdAt: string;
};

export type Transaction = {
  id: string;
  accountId: string;
  type: "debit" | "credit" | "refund";
  amount: number;
  currency: string;
  description: string;
  status: "pending" | "settled" | "declined" | "failed";
  fraudScore?: number;
  createdAt: string;
};

export type Transfer = {
  id: string;
  fromAccountId: string;
  toAccountId: string;
  amount: number;
  currency: string;
  method: "ach" | "wire" | "internal";
  status: "pending" | "completed" | "failed";
  createdAt: string;
};

export type LoanApplication = {
  id: string;
  applicantId: string;
  amount: number;
  term: number;
  purpose: string;
  status: "submitted" | "approved" | "denied" | "disbursed";
  creditScore?: number;
  interestRate?: number;
  createdAt: string;
};

const accounts: Account[] = [
  { id: "acct_001", customerId: "cust_1", type: "checking", balance: 4820.55, currency: "USD", status: "active", createdAt: "2025-01-15T00:00:00Z" },
  { id: "acct_002", customerId: "cust_1", type: "savings", balance: 18200.00, currency: "USD", status: "active", createdAt: "2025-01-15T00:00:00Z" },
  { id: "acct_003", customerId: "cust_2", type: "checking", balance: 1240.80, currency: "USD", status: "active", createdAt: "2025-03-01T00:00:00Z" },
  { id: "acct_004", customerId: "cust_3", type: "credit", balance: -540.00, currency: "USD", status: "frozen", createdAt: "2024-11-10T00:00:00Z" },
];

const transactions: Transaction[] = [
  { id: "txn_001", accountId: "acct_001", type: "debit", amount: 120.00, currency: "USD", description: "Online purchase", status: "settled", fraudScore: 0.03, createdAt: "2026-06-28T10:00:00Z" },
  { id: "txn_002", accountId: "acct_001", type: "credit", amount: 2500.00, currency: "USD", description: "Payroll deposit", status: "settled", createdAt: "2026-06-27T08:00:00Z" },
  { id: "txn_003", accountId: "acct_003", type: "debit", amount: 9800.00, currency: "USD", description: "Wire transfer attempt", status: "declined", fraudScore: 0.87, createdAt: "2026-06-26T15:30:00Z" },
];

const transfers: Transfer[] = [
  { id: "trf_001", fromAccountId: "acct_001", toAccountId: "acct_002", amount: 500.00, currency: "USD", method: "internal", status: "completed", createdAt: "2026-06-25T12:00:00Z" },
];

const loans: LoanApplication[] = [
  { id: "loan_001", applicantId: "cust_1", amount: 15000, term: 36, purpose: "home improvement", status: "approved", creditScore: 740, interestRate: 7.2, createdAt: "2026-06-20T00:00:00Z" },
  { id: "loan_002", applicantId: "cust_3", amount: 5000, term: 24, purpose: "debt consolidation", status: "denied", creditScore: 580, createdAt: "2026-06-22T00:00:00Z" },
];

export function getAccounts(customerId?: string): Account[] {
  if (customerId) return accounts.filter(a => a.customerId === customerId);
  return accounts;
}
export function getAccount(id: string): Account | undefined {
  return accounts.find(a => a.id === id);
}
export function createAccount(customerId: string, type: Account["type"]): Account {
  const acct: Account = { id: `acct_${Date.now()}`, customerId, type, balance: 0, currency: "USD", status: "active", createdAt: new Date().toISOString() };
  accounts.push(acct);
  return acct;
}

export function getTransactions(accountId?: string): Transaction[] {
  if (accountId) return transactions.filter(t => t.accountId === accountId);
  return transactions;
}
export function createTransaction(accountId: string, type: Transaction["type"], amount: number, description: string, fraudScore: number): Transaction {
  const status = fraudScore > 0.75 ? "declined" : "pending";
  const txn: Transaction = { id: `txn_${Date.now()}`, accountId, type, amount, currency: "USD", description, status, fraudScore, createdAt: new Date().toISOString() };
  transactions.push(txn);
  if (status !== "declined") {
    const acct = accounts.find(a => a.id === accountId);
    if (acct) acct.balance += type === "credit" ? amount : -amount;
  }
  return txn;
}
export function settleTransaction(id: string): Transaction | undefined {
  const txn = transactions.find(t => t.id === id);
  if (txn && txn.status === "pending") { txn.status = "settled"; }
  return txn;
}

export function getTransfers(accountId?: string): Transfer[] {
  if (accountId) return transfers.filter(t => t.fromAccountId === accountId || t.toAccountId === accountId);
  return transfers;
}
export function createTransfer(fromAccountId: string, toAccountId: string, amount: number, method: Transfer["method"]): Transfer {
  const trf: Transfer = { id: `trf_${Date.now()}`, fromAccountId, toAccountId, amount, currency: "USD", method, status: "pending", createdAt: new Date().toISOString() };
  transfers.push(trf);
  return trf;
}
export function completeTransfer(id: string): Transfer | undefined {
  const trf = transfers.find(t => t.id === id);
  if (trf) trf.status = "completed";
  return trf;
}

export function getLoans(applicantId?: string): LoanApplication[] {
  if (applicantId) return loans.filter(l => l.applicantId === applicantId);
  return loans;
}
export function applyForLoan(applicantId: string, amount: number, term: number, purpose: string, creditScore: number): LoanApplication {
  const approved = creditScore >= 650 && amount <= 50000;
  const loan: LoanApplication = {
    id: `loan_${Date.now()}`, applicantId, amount, term, purpose,
    status: approved ? "approved" : "denied",
    creditScore,
    interestRate: approved ? Math.max(5.5, 15 - creditScore / 100) : undefined,
    createdAt: new Date().toISOString(),
  };
  loans.push(loan);
  return loan;
}

File: apps/fintech/before/lib/store.py
import json
from datetime import datetime
import uuid

# In-memory storage
accounts = [
    {"id": "a1", "name": "Alice Johnson", "account_type": "checking", "balance": 5234.50, "created_at": "2026-01-15T10:00:00Z"},
    {"id": "a2", "name": "Bob Smith", "account_type": "savings", "balance": 15000.00, "created_at": "2026-02-20T10:00:00Z"},
    {"id": "a3", "name": "Carol White", "account_type": "checking", "balance": 8900.75, "created_at": "2026-03-10T10:00:00Z"},
]

transactions = [
    {"id": "t1", "account_id": "a1", "type": "deposit", "amount": 500.00, "status": "completed", "created_at": "2026-06-28T10:15:00Z"},
    {"id": "t2", "account_id": "a2", "type": "withdrawal", "amount": 200.00, "status": "completed", "created_at": "2026-06-28T11:30:00Z"},
]

transfers = [
    {"id": "xf1", "from_account": "a1", "to_account": "a2", "amount": 100.00, "status": "completed", "created_at": "2026-06-28T10:20:00Z"},
]

loans = [
    {"id": "ln1", "account_id": "a1", "amount": 10000.00, "term_months": 36, "status": "approved", "created_at": "2026-06-28T10:30:00Z"},
]

def get_accounts():
    return accounts

def create_account(name, account_type):
    account = {
        "id": f"a{int(uuid.uuid4().int % 1000000)}",
        "name": name,
        "account_type": account_type,
        "balance": 5000.00,
        "created_at": datetime.utcnow().isoformat() + "Z"
    }
    accounts.append(account)
    return account

def get_account(account_id):
    for a in accounts:
        if a["id"] == account_id:
            return a
    return None

def get_transactions(account_id=None):
    if account_id:
        return [t for t in transactions if t["account_id"] == account_id]
    return transactions

def record_transaction(account_id, transaction_type, amount):
    transaction = {
        "id": f"t{int(uuid.uuid4().int % 1000000)}",
        "account_id": account_id,
        "type": transaction_type,
        "amount": amount,
        "status": "completed",
        "created_at": datetime.utcnow().isoformat() + "Z"
    }
    transactions.append(transaction)
    
    # Update account balance
    account = get_account(account_id)
    if account:
        if transaction_type == "deposit":
            account["balance"] += amount
        else:
            account["balance"] -= amount
    
    return transaction

def get_transfers(account_id=None):
    if account_id:
        return [t for t in transfers if t["from_account"] == account_id or t["to_account"] == account_id]
    return transfers

def initiate_transfer(from_account, to_account, amount):
    transfer = {
        "id": f"xf{int(uuid.uuid4().int % 1000000)}",
        "from_account": from_account,
        "to_account": to_account,
        "amount": amount,
        "status": "initiated",
        "created_at": datetime.utcnow().isoformat() + "Z"
    }
    transfers.append(transfer)
    return transfer

def complete_transfer(from_account, to_account, amount):
    transfer = {
        "id": f"xf{int(uuid.uuid4().int % 1000000)}",
        "from_account": from_account,
        "to_account": to_account,
        "amount": amount,
        "status": "completed",
        "created_at": datetime.utcnow().isoformat() + "Z"
    }
    transfers.append(transfer)
    
    # Update balances
    from_acc = get_account(from_account)
    to_acc = get_account(to_account)
    if from_acc:
        from_acc["balance"] -= amount
    if to_acc:
        to_acc["balance"] += amount
    
    return transfer

def fail_transfer(from_account, to_account, amount):
    transfer = {
        "id": f"xf{int(uuid.uuid4().int % 1000000)}",
        "from_account": from_account,
        "to_account": to_account,
        "amount": amount,
        "status": "failed",
        "created_at": datetime.utcnow().isoformat() + "Z"
    }
    transfers.append(transfer)
    return transfer

def get_loans(account_id=None):
    if account_id:
        return [l for l in loans if l["account_id"] == account_id]
    return loans

def request_loan(account_id, amount, term_months):
    loan = {
        "id": f"ln{int(uuid.uuid4().int % 1000000)}",
        "account_id": account_id,
        "amount": amount,
        "term_months": term_months,
        "status": "pending",
        "created_at": datetime.utcnow().isoformat() + "Z"
    }
    loans.append(loan)
    return loan

def approve_loan(loan_id):
    for l in loans:
        if l["id"] == loan_id:
            l["status"] = "approved"
            return l
    return None

def decline_loan(loan_id):
    for l in loans:
        if l["id"] == loan_id:
            l["status"] = "declined"
            return l
    return None

def get_metrics_formatted():
    from lib.metrics import metrics_store
    result = {}
    for metric, labeled_values in metrics_store.items():
        result[metric] = {}
        for label_key, entry in labeled_values.items():
            labels = json.loads(label_key)
            label_str = ",".join(f"{k}={v}" for k, v in labels.items()) if labels else "total"
            
            if "sum" in entry:
                result[metric][label_str] = {
                    "count": entry["count"],
                    "sum": entry["sum"],
                    "avg": round(entry["sum"] / entry["count"], 2) if entry["count"] > 0 else 0
                }
            else:
                result[metric][label_str] = entry["count"]
    
    return result
